api/localizer/writer.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2019 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+// See proposals/localize-command
+package localizer
+
+import (
+	"net/url"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"sigs.k8s.io/kustomize/api/loader"
+	"sigs.k8s.io/kustomize/kyaml/filesys"
+)
+
+const urlSeparator = "/"
+const localizeDir = "localized-files"
+const alpha = "alpha"
+
+var errNewDirRef = errors.New("target references newDir")
+var errAbsPath = errors.New("absolute paths not yet supported")
+
+func makeAndClean(path string, fSys filesys.FileSystem) (filesys.ConfirmedDir, error) {
+	var cleanPath filesys.ConfirmedDir
+	var err error
+
+	if err = fSys.MkdirAll(path); err != nil {
+		return "", errors.Wrapf(err, filesys.ErrMkDirAll.Error())
+	}
+	if cleanPath, _, err = fSys.CleanedAbs(path); err != nil {
+		return "", errors.Wrapf(err, filesys.ErrCleanAbs.Error())
+	}
+
+	return cleanPath, nil
+}
+
+// NewWriter is the factory method for Writer.
+func NewWriter(target filesys.ConfirmedDir, scope filesys.ConfirmedDir, newDir filesys.ConfirmedDir, fSys filesys.FileSystem) (*Writer, error) {
+	inNewDir, _ := filepath.Rel(scope.String(), target.String())
+	dst, dstErr := makeAndClean(newDir.Join(inNewDir), fSys)
+	if dstErr != nil {
+		return nil, errors.Wrapf(dstErr, "unable to write to newDir")
+	}
+
+	return &Writer{
+		fSys:   fSys,
+		newDir: newDir,
+		root:   target,
+		dstDir: dst,
+	}, nil
+}
+
+// Writer takes care of writing localized files to their appropriate destination
+// in newDir for command kustomize localize. Each Writer is associated with a
+// kustomization root, either remote or local, referenced by target and writes
+// files that the root directly references.
+//
+// Writer is meant to work with loader, which performs the path checks.
+type Writer struct {
+	fSys filesys.FileSystem
+	// newDir argument, as defined by kustomize localize
```

Comment:
Nit: This took me a second to parse. Suggest to either add a newline after each field (between the field and the next comment), or put the comment directly after the field

```
fSys filesys.FileSystem

// newDir ....
```

or 

```
fSys filesys.FileSystem
newDir fileSys.ConfirmedDir // newDir ...
```